### ISSUE TITLE
Layout proposal: nodes "clustered" by groups

### ISSUE THREAD

**Issue by Muntaner** [visjs/vis-network#203]
In various projects where I used VisJS, I often had the need to implement a node clustering feature based on their groups. I'm not talking about the already implemented clustering feature where one node can contain other nodes, but a layout which places nodes of the same group in the same region of space in VisJS canvas.

One good example where you can see what I mean for this is the [world cup example](https://visjs.github.io/vis-network/examples/network/exampleApplications/worldCupPerformance.html) hosted on the site, where all the nodes are nicely clustered and "all nearly placed" based on their group. I also reported this "problem" [in an issue on the original VisJS GitHub project](https://github.com/almende/vis/issues/4055).
But... in the examples, the effect is obtained by the physics engine - as far as I understood - which tends to place highly connected components all together, while I needed a feature where the user simply puts some groups to the nodes and VisJS magically places them in a clustered way.

So, after some experimenting and having fun with VisJS, I came up with a prototypical implementation for this layout, which I'll describe there in a form of simple list of steps.

In short, the algorithm for the feature works by dividing the canvas space where VisJS places the nodes and programmatically assigning xy positions to all the nodes. 

1. The space where VisJS works is considered to be a circle that gets sliced in n partitions, where n is the number of the distinct groups specified in the dataset of nodes. This is simply done by slicing the 360° space for the number of groups with some basic trigonometry.
2. Each of those partitions of space is related to a group of the nodes and has its own center. XY positions of nodes are then calculated in order to be placed on the circumference of such partitions, with the radius of the partition scaled according to how much nodes belong to that group partition (a group of two nodes will get a smaller partition compared to a partition related to a 50 nodes group). Still, this is done by using some trigonometric basic concepts.
3. Physics engine is disabled in order to place the nodes in the xy calculated positions.
4. (optional?) I also used a "fancy" way to visualize the obtained clusters by using a [convex hull](https://en.wikipedia.org/wiki/Convex_hull_algorithms) algorithm. Clusters of node are surrounded within a colored polygon (this is done on the canvas in the _afterDrawing_ function). This can be CPU intensive because convex hull is calculated for each rendering frame for each group, but works pretty well in practice.
5. Result is that different groups of nodes are placed and clustered in circles, placed around the VisJS canvas, amongly spaced, independently from their connections/edges (screenshots attached).

I implemented such prototype in a local minimal project, and I'm attaching some screenshots to show the results.

Problems:
- is this a really needed layout or it is useful only in niche/rare use cases? :) is all of this a not needed overkill?
- design and produce a general version of the algorithm which responds well to highly populated groups and high variance between the population of the groups

**400 nodes, 8 groups**
![400_nodes_8_groups](https://user-images.githubusercontent.com/13050619/68038215-bc57af80-fcc9-11e9-8523-1b83022b4a33.png)
**2 groups clusters**
![2_groups_clusters](https://user-images.githubusercontent.com/13050619/68038218-bcf04600-fcc9-11e9-9392-3b32c1ed348a.png)
**4 groups clusters**
![4_groups_clusters](https://user-images.githubusercontent.com/13050619/68038219-bcf04600-fcc9-11e9-9700-4ca31816e78a.png)
**Convex hull detail**
![convexhull_detail](https://user-images.githubusercontent.com/13050619/68038216-bc57af80-fcc9-11e9-83ef-29fe553e91ed.png)

**Comment by Thomaash** [comment]
I can't really image why someone would do this but that may be just my limited imagination. Could you, please, showcase some projects that good make use of it? Because to me it seems quite weird and the edges go all over the place in a very chaotic manner. Seems really hard to navigate. Though I can't say anything definitive before I actually see it in action and also the code anyway.

In the future I'd like to have the layouts reworked into modular design and custom layout support (something like `network.addLayout("custom-layout-name", customLayoutFunction);`). We could then provide a library with all kinds of niche layouts (like this one) without cluttering the library with code vast majority of people will never use.

**Comment by mojoaxel** [comment]
Hi Michele! Thanks for opening this issue! Please feel free to also share your implementation progress e.g. in a [draft pull-request](https://github.blog/2019-02-14-introducing-draft-pull-requests/).

As @Thomaash already mentioned it would also be useful to describe how these kind of layout could be used.

**Comment by szabi** [comment]
@Muntaner, I'd imagine creating densely connected subgraphs with hidden layers cateters very well for placing them close via the physics engine. You can have physically active (present) but hidden edges.

```js
edges.push({
  from: 1,
  to: 2,
  label: "1--2",
  hidden: true
});
```

Did you experiment with that?

**Comment by clem-41** [comment]
**I actually really, really need such a feature!**
I thought it was already implemented, because I don't really see the point of having groups of nodes if the groups don't "regroup" the nodes in the same area... now I have this gigantic network with nodes from groups scattered all around when I really need them to be in separate areas, because right now you can't see anything, there's way too much information.

You can see my network here: https://clemsmusic.com/thechart/chart_example.php
The nodes are characters, and the groups are shows that these characters belong to. When you click on a node, a popup appears, and if you click on the name of the show in the popup, the nodes from the group (= the characters from the show) appear in red. You can see they're not in the same area at all... which makes it all really hard to read. 

So I vote a big YES to this layout!

**Comment by Thomaash** [comment]
Hi @clem-41,

could you provide some more info about this? Ideally disable physics, position the nodes the way you want and post before and after screenshots. If you want some background highlighting of groups, borders or so sketch it there too (it doesn't have to be pretty or contain a million nodes, just to show what you want so that we're all on the same boat).

Thanks.

PS: If you want to know why it makes sense to have groups of nodes that don't group in the same area look on https://thomaash.github.io/me/#/canvas. There is a group for computers, a group for switches, another group for ports and so on.

**Comment by Thomaash** [comment]
An idea here: We already have groups and clusters. Could we call this districts? Thumbs up? Thumbs down? Better names?

**Comment by clem-41** [comment]
Yes I think districts is a good name for this.

In my specific case, I don't need a precise shape for the districts (like the circles from Muntaner's post). I just want the nodes to stay in the same area, next to the nodes from the other groups.

Here's what my network looks like now: the nodes that belong to a same group aren't next to each other, they're all scattered across the network.
<img width="756" alt="network before group districts" src="https://user-images.githubusercontent.com/56389014/71036605-5199f080-211e-11ea-8f3a-d8ede25b9577.png">

And here's what I would like to get (I only detached the nodes of three groups from the rest of the nodes to show you what it would look like): roughly the same presentation, but with nodes from a group next to the other nodes of the same groups. 
![network after group districts](https://user-images.githubusercontent.com/56389014/71036669-6b3b3800-211e-11ea-80c5-5e02053871a0.jpg)
(Please note that the red borders are just here to show the different groups of nodes, but I don't want them to appear on the network)

**Comment by Thomaash** [comment]
This is mostly brainstorming but:
- A node could belong to a district.
- All nodes in each district would be positioned in isolation.
- All first level districts (districts of nodes) would be positioned.
- All second level districts (districts of districts of nodes) would be positioned.
- All third level districts (districts of districts of districts of nodes) would be positioned.
- And so on.
- All top level districts and nodes not belonging to any district would be positioned.
- Absolute positions would be written into the nodes.

Each district would be done independently in isolation (parallelization opportunity here). This would allow different configurations, layout algorithms etc. for each district. I don't see a difference between a node and a district for layouting (a node has some bounding box and position, a district has some bounding box and position). This could easily be implemented as recursive algorithm running from bottom to top. Basically solving this and  #136 in one go.

The districts could have some basic optional styling:
- borders,
- labels,
- backgrounds.

And someone would have to go, throw away all the algorithms (they're poorly documented and in ancient JS, plus having them in TS would be nice, also the people who wrote them are long gone) and write brand new ones. I think that adapting what we have for such drastic change would be more work though as always I may be wrong.

PS: There's no exact algorithm behind this. If I'm missing something important, please tell.

**Comment by clem-41** [comment]
I'm really not good at this so I can't tell you if something's missing, but what you said sounds good so far! I hope Muntaner comes back here soon so that they can give their opinion on this.

I'm guessing this will take a while to do, right?

**Comment by Thomaash** [comment]
Yeah, this will take a while. Especially since I want to solve a few other issues first.

PS: Contributions are always welcome and deeply appreciated. Just saying in case anyone has some free time.

**Comment by rsshilli** [comment]
This is very useful! I want to add another use case / some more ideas for this feature.  We use vis.js for helping create better/cheaper/safer medications in the pharmaceutical industry.  This example below is test data that's dumbed down a lot.  Imagine here that we're analyzing how to create a better cake:

![image](https://user-images.githubusercontent.com/2487045/71824473-96eb7700-3067-11ea-8d7a-7e48a67b182d.png)

So at the top are the "Final Quality Attributes" of the cake.  This is an analysis of the cake after it's made, like the external color, gluten content, etc.  On the right are the attributes of the incoming materials (think flour, eggs, etc), and on the bottom are the important things to measure at each stage of making the cake.  In this very contrived example, you might mix the dry ingredients separately from the wet ingredients, then mix them all together.  You'll have different mixing times, as well as different measurements of fat content, etc.  This graph helps scientists figure out what they understand, what they don't understand and how it's all related.  The context of which attributes are in which steps of the manufacturing process is very, very helpful. Real-world examples are obviously much, much more complex. 

I've created the above grey boxes listening to the `beforeDrawing` event and drawing the boxes on the canvas.  My users would love to drag and drop those grey boxes around, in order to move the manufacturing steps around and think through different ways to solve various problems.  This feature literally can help design cures to many terrible diseases, including cancer.

**Comment by Thomaash** [comment]
Hi @rsshilli,

thanks for your use case. This is pretty much exactly what I imagine this to look like. It's great to have it on a nice picture and to know there are many people who would use it.

**Comment by meshy** [comment]
This would be great for me too. My use-case for this would be for generating database entity relationship diagrams, where clusters of tables are grouped together into "apps" (in Django terminology).

This would be a particularly nice use for it, as most of the connections would most likely be within the app-groups.

**Comment by Muntaner** [comment]
Hello to everyone, sorry for the delay in the reply.
I'll try to share my thoughts based on the discussion.

@Thomaash: I agree that it is hard to spot the real purpose of such a feature with random data, randomly connected (as I've showcased in the opening comment), but there are some real use cases which need this kind of separation.
Imho, it's needed when you have some strongly interconnected node groups with low inter-group correlation. Apart from the practical cases reported in previous comments, networking and social scenarios come to my mind.
On the naming of the feature, I really like the "district" term!

By the way, [here it is a small repository containing a minimal project with an implementation of the discussed feature](https://github.com/Muntaner/visjs_districts). No tests and all, just some quick and ugly code to make the feature work.
Some comments on the implementation:
- the district disposition of the nodes is applied by using the "computeNodesPositionsDistricts" method in the draw() function, in the attempt of making this feature 100% optional and modular;
- same goes for the convex hull groups rendering, which is contained in the "afterNetworkDrawing" method and is 100% optional;
- in the discussion, I've noticed that the circle disposition of the nodes of a district isn't always a good approach. By the way, this should be easily customizable by changing the "getContourPoint" function logic in districts.js.

I don't know if this is useful or it is the right approach in providing a way to eventually integrate this layout in VisJS code without cluttering the code, but I hope it's a starting point.

**Comment by ccamacho** [comment]
Hi @rsshilli by any chance, do you have an example for rendering the network like you have in https://github.com/visjs/vis-network/issues/203#issuecomment-571164712 ?

Thanks!

**Comment by rsshilli** [comment]
@ccamacho No, unfortunately, I don't. It's quite a bit of code in a private repository.

The gist though is to figure out how to lay everything out and then:
* listen to the "beforeDrawing" event to use raw drawing events to draw the boxes
```
        network.on("beforeDrawing", this.onBeforeGraphDrawing);
```
* list to the "dragStart" event to figure out if one of the boxes is being dragged
```
        network.on("dragStart", this.onDragStart);
```

Then if the user is dragging the box, select the all of the nodes so they're dragging that instead:
```
  onDragStart(event) {
    // Search for a container box that might be clicked on.
    const canvasPosition = event.pointer.canvas;
    const allNodeIdsInContainerBox = this.getAllNodeIdsInContainerBoxAt(canvasPosition);

    if (allNodeIdsInContainerBox) {
      window.network.selectNodes(allNodeIdsInContainerBox);
    }
  }
```

As they drag the nodes around now, the box will still keep getting drawn using the `beforeDrawing` event as they drag the items around the screen.

Hopefully, that helps.

**Comment by chichistrong** [comment]
I totally need this for my project. I am currently working on a dependency mapping tool and I want the user to be able to arrange dependencies by it’s group, if they want to. Just like in your example layout. I am surprised nobody has ever thought of adding this to vis us before, but goodness I need this.

**Comment by nlinton** [comment]
This would also be really useful for epidemiology, where individual cases are connected to some other cases, but then also might be connected to (or only connected to) a "cluster" (i.e. a location or event associated with multiple cases).

**Comment by monicaraghuwanshi** [comment]
I also have a similar usecase. Would be great if it is possible.

**Comment by Muntaner** [comment]
> @Muntaner Great work! I believe this is a long-waited feature for some vis.js users. Is there a code repo?

I used to have a repo for this feature, but I realized I really don't have the time to dedicate to the project in order to polish it and properly find a way to integrate it with the VisJS codebase and I decided to close it.

All the code (raw, rough, untested, whatever :P ) that produces the districts we discussed in this ticket is available in the codebox I linked in the previous comment. It must not be considered at all production-ready code, but a quick way to implement the feature on top of the existing VisJS release - no need to hijack in the original code or patch it. 
It would be great if anyone with more time and dedication could start from that dirty implementation and improve it, and eventually understand if it can be someway integrated with the existing library.

**Comment by ooker777** [comment]
# Another example case
For a directed graph there will be source point, which have the in-degree=0. I want to group all nodes that are descendants of a source point together, and draw a border outside of them. So a graph with many source points will have many components like that. And if a node is a descendant of many source points, then it will be in the intersect of many components.

This is also asked in Stack Overflow: [Bordered group of nodes in a network graph?](https://stackoverflow.com/q/26703717/3416774)

**Comment by Muntaner** [comment]
> # Another example case
> For a directed graph there will be source points, which have the in-degree=0. I want to group all nodes that are descendants of a source point together, and draw a border outside of them. So a graph with many source points will have many components like that. And if a node is a descendant of many source points, then it will be in the intersect of many components.
> 
> This is also asked in Stack Overflow: [Bordered group of nodes in a network graph?](https://stackoverflow.com/q/26703717/3416774)

This proposal would probably solve the "group all descendants together", but you would still need some tweaks and customization to correctly place the source points and to correctly implement the intersection part.

**Comment by xmedeko** [comment]
Our use case is: we have (about 10-200) projects (vertices) with dependencies (edges). Projects are organized in (about 3-10) groups. We want to show and edit project dependencies (edges). However the projects - vertices should be in a "group" bubble (rectangle), so that a user can find it quickly.

Maybe this feature could be called "opened cluster"?
